zebra-state/src/service/write.rs

```diff
@@ -0,0 +1,138 @@
+//! Writing blocks to the finalized and non-finalized states.
+
+use std::sync::{Arc, Mutex};
+
+use zebra_chain::block::{self, Height};
+
+use crate::service::{
+    finalized_state::FinalizedState,
+    queued_blocks::{QueuedFinalized, QueuedNonFinalized},
+    ChainTipBlock, ChainTipSender,
+};
+
+/// Reads blocks from the channels, writes them to the `finalized_state`,
+/// and updates the `chain_tip_sender`.
+///
+/// TODO: pass the non-finalized state and associated update channel to this function
+#[instrument(skip(
+    finalized_block_write_receiver,
+    non_finalized_block_write_receiver,
+    invalid_block_reset_sender,
+    chain_tip_sender
+))]
+pub fn write_blocks_from_channels(
+    mut finalized_block_write_receiver: tokio::sync::mpsc::UnboundedReceiver<QueuedFinalized>,
+    mut non_finalized_block_write_receiver: tokio::sync::mpsc::UnboundedReceiver<
+        QueuedNonFinalized,
+    >,
+    mut finalized_state: FinalizedState,
+    invalid_block_reset_sender: tokio::sync::mpsc::UnboundedSender<block::Hash>,
+    chain_tip_sender: Arc<Mutex<ChainTipSender>>,
+) {
+    // Write all the finalized blocks sent by the state,
+    // until the state closes the finalized block channel's sender.
+    while let Some(ordered_block) = finalized_block_write_receiver.blocking_recv() {
+        // TODO: split these checks into separate functions
+
+        if invalid_block_reset_sender.is_closed() {
+            info!("StateService closed the block reset channel. Is Zebra shutting down?");
+            return;
+        }
+
+        // Discard any children of invalid blocks in the channel
+        //
+        // `commit_finalized()` requires blocks in height order.
+        // So if there has been a block commit error,
+        // we need to drop all the descendants of that block,
+        // until we receive a block at the required next height.
+        let next_valid_height = finalized_state
+            .db
+            .finalized_tip_height()
+            .map(|height| (height + 1).expect("committed heights are valid"))
+            .unwrap_or(Height(0));
+
+        if ordered_block.0.height > next_valid_height {
```

Comment:
Could it happen that we attempt to commit a block with a height lower than `next_valid_height`? In that case, we would need
```suggestion
        if ordered_block.0.height != next_valid_height {
```

Reply:
You're right, we should use the same check as `commit_finalized()` does.